src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourceMappingOptionsControl.xaml.cs

```diff
@@ -0,0 +1,216 @@
+                {
+                    packageSourceMappingProvider.SavePackageSourceMappings(packageSourceMappingsSourceItems);
+                }
+            }
+            // Thrown during creating or saving NuGet.Config.
+            catch (NuGetConfigurationException ex)
+            {
+                MessageHelper.ShowErrorMessage(ex.Message, Resx.ErrorDialogBoxTitle);
+                return false;
+            }
+            // Thrown if no nuget.config found.
+            catch (InvalidOperationException ex)
+            {
+                MessageHelper.ShowErrorMessage(ex.Message, Resx.ErrorDialogBoxTitle);
+                return false;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                MessageHelper.ShowErrorMessage(Resx.ShowError_ConfigUnauthorizedAccess, Resx.ErrorDialogBoxTitle);
+                return false;
+            }
+            // Unknown exception.
+            catch (Exception ex)
+            {
+                MessageHelper.ShowErrorMessage(Resx.ShowError_ApplySettingFailed, Resx.ErrorDialogBoxTitle);
+                ActivityLog.LogError(NuGetUI.LogEntrySource, ex.ToString());
+                return false;
+            }
+            return true;
+        }
+
+        // Returns true if there are changes between existingSourceMappings and packageSourceMappings.
+        private static bool SourceMappingsChanged(IReadOnlyList<PackageSourceMappingSourceItem> existingSourceMappings, IReadOnlyList<PackageSourceMappingSourceItem> packageSourceMappings)
+        {
+            if (existingSourceMappings.Count != packageSourceMappings.Count)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < existingSourceMappings.Count; ++i)
+            {
+                //checks to see if keys match
+                if (existingSourceMappings[i] != packageSourceMappings[i])
+                {
+                    return true;
+                }
+                //checks to see if all patterns match
+                if (existingSourceMappings[i].Patterns.Count != packageSourceMappings[i].Patterns.Count)
+                {
+                    return true;
+                }
+                for (int j = 0; j < existingSourceMappings[i].Patterns.Count; ++j)
+                {
+                    if (existingSourceMappings[i].Patterns[j] != packageSourceMappings[i].Patterns[j])
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private IReadOnlyList<SourceMappingViewModel> CreateViewModels(IReadOnlyList<PackageSourceMappingSourceItem> originalMappings)
+        {
+            var uiSourceMappings = new Dictionary<string, List<PackageSourceContextInfo>>();
+            foreach (PackageSourceMappingSourceItem sourceItem in originalMappings)
```

Comment:
I think we're missing listening to config changes changes and refreshing the values when the config does change. 

An example is adding a source.
![repro](https://user-images.githubusercontent.com/2878341/195225338-fccb8b91-cc54-4512-a3d6-acf773c72b4d.gif)

Reply:
> "listening to config changes" sounds like you're suggesting that adding a package source should immediately write to the config while the dialog is still open, is that correct? 

Not necessarily. I think it was a leap on mind to suggest solution. 

Let's focus on the scenario. 

- Add a source through the options UI.
- Go and add mappings for that source *without* having to reopen the the options UI *should work*. 

> If we do this, then the commit behavior that VS Options OK buttons typically provide will be pointless, as the config was already modified. Will customers expect this behavior change? I believe most pages commit-on-OK.

I rephrased my suggestion. I should have merely pointed out the user scenario, rather than offer fixes or comment on technical limitations.

I think we should fix the scenario I pointed out. It doesn't have to be done in this PR, it can be done in a future one, but imo it should be done before release.

Reply:
> It doesn't have to be done in this PR, it can be done in a future one,

Issue created and I'll evaluate it next sprint (and look for customer input) (https://github.com/NuGet/Home/issues/12155). 

> Go and add mappings for that source without having to reopen the options UI should work.

...will require some care, since likely it will require an implementation involving:

> I could keep new sources in memory, just in case they want to immediately map it.

Reply:
> Issue created and I'll evaluate it next sprint (and look for customer input) (https://github.com/NuGet/Home/issues/12155).

Makes sense. I don't think we should wait for customer input. 
I think this is a core scenario. 

We'll review it during one of triage meetings and we can go from there.